jetty-util/src/main/java/org/eclipse/jetty/util/PidFile.java

```diff
@@ -0,0 +1,78 @@
+//
+// ========================================================================
+// Copyright (c) 1995 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import org.eclipse.jetty.util.annotation.Name;
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.eclipse.jetty.util.component.LifeCycle;
+import org.eclipse.jetty.util.thread.ShutdownThread;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.nio.file.StandardOpenOption.CREATE;
+import static java.nio.file.StandardOpenOption.TRUNCATE_EXISTING;
+import static java.nio.file.StandardOpenOption.WRITE;
+
+/**
+ * Create a PID file for the running process.
+ *
+ * <p>
+ *     This class will automatically register itself with a call to
+ *     {@link ShutdownThread#register(LifeCycle...)}.
+ * </p>
+ */
+public class PidFile extends AbstractLifeCycle
+{
+    private static final Logger LOG = LoggerFactory.getLogger(PidFile.class);
+
+    private final Path pidFile;
+
+    public PidFile(@Name("file") String filename)
+    {
+        pidFile = Paths.get(filename).toAbsolutePath();
+        try
+        {
+            // Create the PID file as soon as possible.
+            // We don't want for doStart() as we want the PID creation to occur quickly for jetty.sh
+            long pid = ProcessHandle.current().pid();
+            Files.writeString(pidFile, Long.toString(pid), UTF_8, CREATE, WRITE, TRUNCATE_EXISTING);
+            if (LOG.isDebugEnabled())
+                LOG.debug("PID File: {}", pidFile);
+            ShutdownThread.register(this);
+        }
+        catch (Throwable t)
+        {
+            LOG.warn("Unable to create pidFile: {}", pidFile, t);
+        }
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        ShutdownThread.deregister(this);
+        try
+        {
+            Files.deleteIfExists(pidFile);
+        }
+        catch (Throwable t)
+        {
+            LOG.warn("Unable to remove pidFile: {}", pidFile, t);
+        }
```

Comment:
@joakime this doesn't work. If we stop a server before exiting the JVM, we still want the PID file deleted.

Reply:
... or the other approach is to never delete the pid file.   That way we are safe for any lockups between the deletion and the actual file exit.     Code that deals with PID files should be robust for processes that have already exited....  but what is the re-use policy of PIDs?

Reply:
> @joakime this doesn't work. If we stop a server before exiting the JVM, we still want the PID file deleted.

That would be invalid, as the process that PID is pointing to still exists in that scenario.
The PID refers to the system process, not the Server.
Once the system process (the JVM) exits (or is exiting), then we should attempt to remove the PID.

If the PID exists when we start again, that would not impact service `start`, but it would impact service `status` style calls.
The following is how systemd `status` calls handle the PID file from a `status` point of view.

Exit Code | Meaning
---- | ----
0 | PID exists, system process exists, status reported from service itself
1 | PID exists, system process does not exist (no status from service) service is reported as "not running"
3 | PID does not exist, service is reported as "not running"
4 | PID exists, system process exists, status not able to be determined from service (hung process, no reply, etc), reported as "unable to determine status", with PID displayed in output.

_Exit code 2 is not used in systemd status calls AFAICT_

Reply:
@Joakim we want the file deleted when the jvm exits regardless of the server is stopped or started. Removing the shutdown hook is wrong. Just don't register more than 1 per filename.

Reply:
@gregw Just a heads up, you're mentioning the wrong Joakim :) It has happened a few times in this repo.

My namesake has an "e" at the end.

Reply:
> @joakim we want the file deleted when the jvm exits regardless of the server is stopped or started. Removing the shutdown hook is wrong. Just don't register more than 1 per filename.

I don't understand.

This is never added to a Server, or any other Jetty Container.
This only exists as a `LifeCycle` to satisfy the Jetty `org.eclipse.jetty.util.thread.ShutdownThread` mechanism.

To recap.

The `org.eclipse.jetty.util.thread.ShutdownThread` ...
1. Is a singleton
2. Is _not_ a Jetty `org.eclipse.jetty.util.component.LifeCycle` or `org.eclipse.jetty.util.component.Container` class.
3. Has no LifeCycle of its own (start/stop).
4. Is registered as a JVM Runtime Shutdown Thread - `Runtime.getRuntime().addShutdownHook(ShutdownThread)` (once and only once)
5. Collects a `List<LifeCycle>` of registered LifeCycle's to call when the JVM shutdown occurs. (because of a call to `ShutdownThread.register(LifeCycle)`)
6. When JVM shutdown occurs the `ShutdownThread.run()` executes.
7. All registered LifeCycle's (on the `ShutdownThread`) have their `LifeCycle.stop()` method called (and optionally a `Destroyable.destroy()`) methods.

I feel the fact that this is a `LifeCycle` is confusing things for you.
The `PidFile` exists only for `ShutdownThread` use in this current HEAD (it is never started, or managed, or added to another lifecycle container).
The `PidFile.doStop()` method is what is executed during JVM shutdown.

Options to get unstuck from this apparent confusion over `LifeCycle` and `ShutdownThread`.

### (A) Remove `LifeCycle` and `ShutdownThread` use entirely in `PidFile`.

Just make `PidFile` a `Runnable`, and have it register itself as a `Runtime.getRuntime().addShutdownHook(PidFile)`.

### (B) Move `ShutdownThread.register()` to `jetty-pid.xml`

`PidFile` remains a `LifeCycle` class, but nothing in Jetty code or an XML adds it to any Jetty `org.eclipse.jetty.util.component.Container`.
Leave that strange use case for embedded jetty users.
Have `jetty-pid.xml` create the `PidFile` instance and register it with the `ShutdownThread`.
Never unregister, as it doesn't make sense for this use case.

Reply:
I like `A` better because seems way simpler.